kep/115-Subgroup-support/README.md

```diff
@@ -0,0 +1,208 @@
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### LeaderWorkerSet API
+We extend the LeaderWorkerSet API to introduce a new field: subGroupSize to opt in and set the number of pods that each subgroup will contain. Current behavior is kept if not set. 
+
+```
+type LeaderWorkerSetSpec struct {
+	 // Number of pods per subgroup. This value is immutable,
+   // and must not be greater than LeaderWorkerSet.Spec.Size.
+	SubGroupSize *int32 'json:"subGroupSize,omitempty'
+} 
+```
+
+### Exclusive Topology Support
+LeaderWorkerSet can guarantee that the leader and the workers are placed in the same topology if the `leaderworkerset.gke.io/exclusive-topology` annotation is set. Similarly, we will support that the pods within the same subgroup will be placed in the same topology with a new annotation `leaderworkerset.gke.io/exclusive-topology-subgroup`, so that the new changes can support up to two levels of pod affinity. 
+
+### Implementation
+- Two new annotations will be added
+  - `leaderworkerset.gke.io/exclusive-topology-subgroup `
+  - `leaderworkerset.gke.io/subgroup-size` 
+- Three new labels will be added,
+  - `leaderworkerset.gke.io/subgroup-index = worker-index/subGroupSize`
+    - Tracks which subgroup the pod is part of 
+  - `leaderworkerset.gke.io/subgroup-worker-index = worker-index%subGroupSize`
+    - index/identity of a pod inside the pod's subgroup
+  - `leaderworkerset.gke.io/subgroup-key` 
+    - Pods that are part of the same subgroup will have an annotation that is a unique hash value will be generated from the name of the leader, and the subgroup-index
+
+To support exclusive placement at the subgroup level, the pod webhook will inject the new labels, and set the pod affinity/anti-affinity on all the pods. If both levels of pod affinity are set, then the leader pod will contain two pod affinities, while the workers will have a node selector, and a single pod affinity set. 
+
+### TPU Environment Variable Injection
+Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WORKER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_HOSTNAMES` will only be a list of the pods in the same subgroup.
+
+LeaderWorkerSet supports the leader not requesting TPU resources. This raises a problem when determining the values of `subgroup-index` and `subgroup-worker-index`. Suppose we have two TPU slices with two hosts each. Since the leader doesn’t request TPU resources, there will be four workers + the leader, meaning that one of the workers will have a worker index of four. Because of the way the new labels are calculated, this worker will have a subgroup-index of two, creating three subgroup indices (0,1,2) even though there are only two TPU slices.
+
+If the leader does not request TPU resources, then the labels will have the following values
```

Comment:
Will this lead to vendor problem, because the logic is special as normal. Hope not here.

Reply:
By vendor problem you mean the logic changes based on whether or not the user uses TPUs?

Reply:
I mean we should not change the label values because of different scenarios, they're identifiers, but  we can leave the logics to the injecting part as we do today. I guess it's possible.

Reply:
This is the last big concern as leaving the TPU logic to accelerators as we do today. see https://github.com/kubernetes-sigs/lws/blob/dde4a46b1db6ce431449ef19c0566000d878ba6e/pkg/utils/accelerators/tpu.go#L105